### PR TITLE
ERC721 ownership precompiles

### PR DESCRIPTION
**Pull request by JasonTulp**
Adding token approval module to handle ERC721 token approvals

### CODE REVIEW

File: runtime/src/lib.rs [hunk 1]
```diff
@@ -633,6 +633,12 @@ impl crml_eth_bridge::Config for Runtime {
 	type FinalSessionTracker = Staking;
 }
 
+impl crml_token_approvals::Config for Runtime {
+	type Event = Event;
+	type MultiCurrency = GenericAsset;
+	type IsTokenOwner = Nft;
```
Comment (jordy25519) [inline runtime/src/lib.rs]: 🔥

File: primitives/src/types.rs [hunk 2]
```diff
@@ -71,6 +71,18 @@ pub type Block = generic::Block<Header, OpaqueExtrinsic>;
...
+/// Uniquely identifies a collection
+pub type CollectionId = u32;
+
+/// Auto-incrementing Uint
+/// Uniquely identifies a series of tokens within a collection
+pub type SeriesId = u32;
+
+/// Auto-incrementing Uint
+/// Uniquely identifies a token within a series
+pub type SerialNumber = u32;
```
Comment (jordy25519) [inline primitives/src/types.rs:84]: could you import these into nft/types and make sure they're used there

Comment (jordy25519) [reply primitives/src/types.rs:84]: nvm you did it already :)

File: crml/nft/rpc/runtime-api/Cargo.toml [hunk 3]
```diff
@@ -11,6 +11,7 @@ sp-api = { git = "https://github.com/cennznet/substrate", rev = "2db18f864ae2bb5
 sp-std = { git = "https://github.com/cennznet/substrate", rev = "2db18f864ae2bb50e60e497b7ac2c521daf83af6", default-features = false }
 sp-runtime = { git = "https://github.com/cennznet/substrate", rev = "2db18f864ae2bb50e60e497b7ac2c521daf83af6", default-features = false }
 crml-nft = { path = "../../", default-features = false }
+cennznet-primitives = { path = "../../../../primitives" }
```
Comment (jordy25519) [inline crml/nft/rpc/runtime-api/Cargo.toml]: `default-features = false` here

File: primitives/src/types.rs (hunk 2 above)
Comment (JasonTulp) [reply primitives/src/types.rs:84]: I think they already are imported into nft/types

File: evm-precompiles/erc721/src/lib.rs [hunk 4]
```diff
@@ -269,11 +269,13 @@ where
 		}
 		let serial_number: SerialNumber = serial_number.saturated_into();
 		let token_id = (series_id_parts.0, series_id_parts.1, serial_number);
-		let approved_account: Runtime::AccountId = crml_token_approvals::Module::<Runtime>::erc721_approvals(token_id);
+		let approved_account: H160 = Runtime::AddressMapping::from_account_id(
+			crml_token_approvals::Module::<Runtime>::erc721_approvals(token_id),
```
Comment (jordy25519) [inline evm-precompiles/erc721/src/lib.rs:277]: should add this extra storage read cost to the gasmeter

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -97,6 +97,9 @@ use pallet_evm::{Account as EVMAccount, EnsureAddressTruncated, FeeCalculator, R
 pub mod constants;
 use constants::{currency::*, time::*};
 
+#[cfg(test)]
+pub use crate::tests::common::mock::ExtBuilder;
```
Comment (jordy25519) [inline runtime/src/lib.rs]: still need this?

File: evm-precompiles/erc721/src/lib.rs [hunk 6]
```diff
@@ -274,23 +268,76 @@ where
...
-			// TODO: use approval stuff..
+			return Err(error("caller not approved").into());
+		}
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(true).build(),
+			logs: Default::default(),
```
Comment (jordy25519) [inline evm-precompiles/erc721/src/lib.rs]: need to add the `Transfer` log for this operation to match erc721 and allow dapps to distinguish between normal transfer and approval transfer

File: crml/token-approvals/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+		) -> DispatchResult {
+			// mapping(uint256 => address) private _tokenApprovals;
+
+			let _ = ensure_none(origin)?;
+			ensure!(caller != operator_account, Error::<T>::CallerNotOperator);
+			// Check that origin owns NFT
+			ensure!(T::IsTokenOwner::check_ownership(&caller, &token_id), Error::<T>::NotTokenOwner);
+			ERC721Approvals::<T>::insert(token_id, operator_account.clone());
+
+			Self::deposit_event(RawEvent::NFTApprovalSet(Some(operator_account), token_id));
```
Comment (jordy25519) [inline crml/token-approvals/src/lib.rs]: let's log the relevant event in the precompile so it's emmited for EVM based apps

File: crml/token-approvals/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	}
+}
+
+impl<T: Config> Module<T> {
+	/// Removes the approval of a single NFT
+	/// Triggered by transferring the token
+	pub fn remove_erc721_approval(token_id: &TokenId) {
+		// Check that origin owns NFT
+		ERC721Approvals::<T>::remove(token_id);
+		Self::deposit_event(RawEvent::NFTApprovalSet(None, *token_id));
```
Comment (jordy25519) [inline crml/token-approvals/src/lib.rs]: same here, any erc721 events should be logged via the evm

**Review by jordy25519 — COMMENTED** [review]
test look good to me.
need to emit logs within the EVM/precompiles

File: runtime/tests/evm_precompiles.rs [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+			Nft::token_owner((collection_id, series_id), serial_number),
+			token_owner.clone()
+		);
+
+		let address: H160 = Runtime::runtime_id_to_evm_id((collection_id, series_id)).into();
+		let context: Context = Context {
+			address,
+			caller: token_owner_eth,
+			apparent_value: U256::default(),
+		};
```
Comment (jordy25519) [inline runtime/tests/evm_precompiles.rs]: nit: some of this setup is very repetitive, easier to see the important part of the test if this is moved to a setup function

File: evm-precompiles/erc721/src/lib.rs [hunk 10]
```diff
@@ -274,23 +271,84 @@ where
 			return Err(error("expected token id <= 2^32").into());
 		}
 		let serial_number: SerialNumber = serial_number.saturated_into();
+		let token_id = (series_id_parts.0, series_id_parts.1, serial_number);
+		let approved_account: H160 = Runtime::AddressMapping::from_account_id(
+			crml_token_approvals::Module::<Runtime>::erc721_approvals(token_id),
+		);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
```
Comment (jordy25519) [inline evm-precompiles/erc721/src/lib.rs]: nit: with gas you should always charge the cost first

File: crml/support/src/lib.rs [hunk 11]
```diff
@@ -60,6 +62,27 @@ where
 	}
 }
 
+impl<AccountId> AddressMappingReversibleExt<AccountId> for PrefixedAddressMapping<AccountId>
+where
+	AccountId: From<[u8; 32]> + Into<[u8; 32]>,
+{
+	fn from_account_id(address: AccountId) -> H160 {
```
Comment (jordy25519) [inline crml/support/src/lib.rs:69]: should add a test for this

**Review by jordy25519 — APPROVED** [review]
lgtm - couple last things

File: evm-precompiles/erc721/Cargo.toml [hunk 12]
```diff
@@ -18,19 +19,26 @@ sp-core = { git = "https://github.com/cennznet/substrate", rev = "2db18f864ae2bb
 sp-runtime = { git = "https://github.com/cennznet/substrate", rev = "2db18f864ae2bb50e60e497b7ac2c521daf83af6", default-features = false, version = "4.0.0-dev" }
 sp-std = { git = "https://github.com/cennznet/substrate", rev = "2db18f864ae2bb50e60e497b7ac2c521daf83af6", default-features = false, version = "4.0.0-dev" }
 num_enum = { version = "0.5.3", default-features = false }
+cennznet-primitives = { path = "../../primitives", default-features = false }
+
+[dev-dependencies]
+cennznet-runtime = { path = "../../runtime" }
```
Comment (jordy25519) [inline evm-precompiles/erc721/Cargo.toml]: this can go to

**Pull request merged**